apps/examples/pm/pm_test.c

```diff
@@ -0,0 +1,183 @@
+
+/**************************************************************************
+* Private Variables
+**************************************************************************/
+
+/**************************************************************************
+* Private Functions
+**************************************************************************/
+
+/**************************************************************************
+* Public Functions
+**************************************************************************/
+
+/**************************************************************************
+* Local Functions
+**************************************************************************/
+
+void pmtest_dev_notify(struct pm_callback_s *cb, int domain, int state)
+{
+	printf("Recevied notify call for %s device in %d domain for %d state \n", cb->name, domain, state);
+}
+
+int pmtest_dev_prepare(struct pm_callback_s *cb, int domain, int state)
+{
+	printf("Recevied prepare call for %s device in %d domain for %d state \n", cb->name, domain, state);
+	return OK;
+}
+
+/**
+* @fn                   :pm_test_kthread
+* @brief                :This tc tests the different  power states provided
+*                        by the PM.
+* @Scenario             :The function shall tests the power states provided
+*                        by PM by sleeping for a certain period & recording
+*                        the sleeping time in each sleep state.
+* @API'scovered         :pm_changestate(), pm_dumpstates(), pm_get_domainmetrics()
+* @Preconditions        :Device should have entered different sleep states
+* @Postconditions       :none
+* @return               :void
+*/
+static int pm_test_kthread(int argc, char *argv[])
+{
+	int pm_state = PM_TEST_STATE;
+	int ret = -1;
+	struct timespec start_time;
+	struct timespec elapsed_time;
+
+	clock_gettime(CLOCK_REALTIME, &start_time);
+	elapsed_time.tv_sec = start_time.tv_sec;
+
+	while (elapsed_time.tv_sec <= (start_time.tv_sec + PM_TEST_DURATION_IN_SECS)) {
+		sleep(PM_TEST_SLEEP_TIME);
+		ret = pm_changestate(PM_TEST_DOMAIN, pm_state);
+		pm_dumpstates();
+#ifdef CONFIG_PM_METRICS
+		int s = 0;
+		pm_get_domainmetrics(0);
+		for (s = 0; s < CONFIG_PM_NSTATE; s++) {
+			printf("Time spent in state%d :%d\n", s, pm_time_in_each_s[s]);
+		}
+#endif
+		clock_gettime(CLOCK_REALTIME, &elapsed_time);
+	}
+
+	/* Return to normal state */
+	pm_state = PM_NORMAL;
+	ret = pm_changestate(PM_TEST_DOMAIN, pm_state);
+
+	return 0;
+}
+
+/* Launch pm test thread */
+
+void pm_test_launch_kthread(void)
+{
+	int i = 0;
+	pthread_t pid;
+
+	if (pthread_create(&pid, NULL, (pthread_addr_t)pm_test_kthread, (pthread_addr_t)NULL) < 0) {
+		printf("pmtest kthread launch failed\n");
+		for (i = 0; i < PM_TEST_DEVICES; i++) {
+			pm_unregister(0, &pmtest_cbarray[i]);
+		}
+	}
+
+	pthread_join(pid, NULL);
+}
+
+/****************************************************************************
+ * PM_test_main
```

Comment:
When Benix made this, I asked the reason why you added it into pm inside, not apps.
He said, no way to call pm internal APIs.
In this file, there are pm internal APIs and making kthread.
It is not correct.
That's why you guys made testcase driver.

Reply:
have added the PM module test to testcase driver.